sdk/cosmos/azure-cosmos/azure/cosmos/_read_items_helper.py

```diff
@@ -0,0 +1,261 @@
+
+        :param items_by_partition: A dictionary mapping partition key range IDs to lists of items with indices.
+        :type items_by_partition: dict[str, list[tuple[int, str, "_PartitionKeyType"]]]
+        :return: A list of query chunks, where each chunk is a dictionary with a single partition.
+        :rtype: list[dict[str, list[tuple[int, str, "_PartitionKeyType"]]]]
+        """
+        query_chunks = []
+        for partition_id, partition_items in items_by_partition.items():
+            # Split large partitions into chunks of self.max_items_per_query
+            for i in range(0, len(partition_items), self.max_items_per_query):
+                chunk = partition_items[i:i + self.max_items_per_query]
+                query_chunks.append({partition_id: chunk})
+        return query_chunks
+
+    def _execute_query_chunk_worker(
+            self, partition_id: str, chunk_partition_items: List[Tuple[int, str, "_PartitionKeyType"]]
+    ) -> Tuple[List[Tuple[int, Dict[str, Any]]], float]:
+        """Synchronous worker to build and execute a query for a chunk of items.
+
+        :param str partition_id: The ID of the partition to query.
+        :param list[tuple[int, str, any]] chunk_partition_items: A chunk of items to be queried.
+        :return: A tuple containing the list of query results with original indices and the request charge.
+        :rtype: tuple[list[tuple[int, dict[str, any]]], float]
+        """
+        captured_headers = {}
+
+        # Create a local hook that just captures headers
+        def local_response_hook(hook_headers, _):
+            # Only capture headers, don't call original hook here
+            captured_headers.update(hook_headers)
+
+        request_kwargs = self.kwargs.copy()
+        request_kwargs['response_hook'] = local_response_hook
+
+        # Create a map of original indices to document IDs for lookup after query
+        id_to_idx = {item[1]: item[0] for item in chunk_partition_items}
+        # Strip the index for the query builder
+        items_for_query = [(item[1], item[2]) for item in chunk_partition_items]
+
+        if _QueryBuilder.is_id_partition_key_query(items_for_query, self.partition_key_definition):
+            query_obj = _QueryBuilder.build_id_in_query(items_for_query)
+        elif _QueryBuilder.is_single_logical_partition_query(items_for_query):
+            query_obj = _QueryBuilder.build_pk_and_id_in_query(items_for_query, self.partition_key_definition)
+        else:
+            # Pass only the current chunk to the query builder
+            partition_items_dict = {partition_id: items_for_query}
+            query_obj = _QueryBuilder.build_parameterized_query_for_items(
+                partition_items_dict, self.partition_key_definition)
+
+        # Execute query with the response hook
+        query_iterator = self.client.QueryItems(self.collection_link, query_obj, self.options, **request_kwargs)
```

Comment:
query is not always cheap than point read, for partition with only 1 item, using point read can be cheaper